File: src/m33_farrsc.c
/* m33_farrsc.c -- a program whose resource does not fit the pool.
 *
 * FARRSC.RSC is 42 KB, 702 objects in 28 trees (tools/farrsc.py); the
 * application pool is 14 KB of bank $00.  Until docs/far-trees.md the
 * AES could not load it at all.  Now it can, for a program that says it
 * can take a far address -- and this is that program, compiled
 * --data-model=large, whose kit sets the word in rsrc_load's int_in that
 * asks for one (src/app/gemlib.c).  The SAME SOURCE is also built
 * --data-model=small as M33S.G4A, whose kit passes no such word: for it
 * rsrc_load must return 0 and nothing else may happen, because a program
 * holding 16-bit pointers handed a far resource would get them silently
 * truncated.  test-m33 runs both.
 *
 * What the large build proves, in order, each step published NEAR so the
 * gate can read it by symbol against the address the loader reported
 * (the m29 pattern):
 *
 *   m33_loaded    rsrc_load returned 1
 *   m33_bank      the bank of tree 0's address from rsrc_gaddr -- NOT ZERO
 *                 is the whole point; a pool resource would answer $00
 *   m33_hdrhi     global[8], the high word of the header's address: the
 *                 AES's own account of where it put the file
 *   m33_ap6       global[6], ap_ptree's high word, likewise
 *   the dialog is CENTRED AND DRAWN from the far tree (the gate looks at
 *                 the screen while this program waits for a key)
 *   m33_findok    objc_find at a point inside the OK button answers FR_OK
 *                 -- a walk through 27 far objects that lands on the
 *                 right one
 *   m33_str0      the first byte of free string 0, read through the far
 *                 address rsrc_gaddr(R_STRING) handed back: 'F'
 *   m33_mtfar     menu_text WROTE WHERE ob_spec POINTS, which for this
 *   m33_mtnear    resource is far memory -- and did NOT write to bank $00
 *                 at the address ob_spec truncates to.  This pair is here
 *                 because mn_text did exactly that until 2026-09-18: it
 *                 took `(char *)(uint16_t)ob_spec` and dropped the bank,
 *                 and since far_alloc hands out whole banks the remainder
 *                 is the string's OFFSET IN THE .RSC FILE.  QED's
 *                 "  Makefile..." lives at offset $08C4, so the copy
 *                 landed on $0008C4 -- inside SpartaDOS X -- and the next
 *                 directory read jumped into a BRK.  The failure showed up
 *                 as a file selector that drew and then hung the machine,
 *                 with nothing to connect it to a menu call, so what is
 *                 checked here is the WRITE, not any symptom of it.
 *   m33_gfree     rsrc_free returned 1, the far block given back
 */
#include "portab.h"
#include "gem.h"
#include "farrsc.h"

NEAR WORD m33_step;           /* how far it got, if it did not get to the end */
NEAR WORD m33_loaded;
NEAR WORD m33_bank, m33_lo;   /* tree 0's address, high and low */
NEAR WORD m33_hdrlo, m33_hdrhi;
NEAR WORD m33_ap5, m33_ap6;
NEAR WORD m33_cx, m33_cy, m33_cw, m33_ch;   /* where form_center put it */
NEAR WORD m33_find, m33_findok;
NEAR WORD m33_str0;
NEAR WORD m33_gfree;
NEAR WORD m33_model;          /* 4 = large, 2 = small: sizeof a pointer */
NEAR WORD m33_mtfar;          /* menu_text wrote at the far ob_spec */
NEAR WORD m33_mtnear;         /* ...and bank $00 at its low word is untouched */
NEAR WORD m33_mtlo;           /* that low word, for the gate to report */

/* The name near, so that neither kit has to bounce it: what this program
 * tests is the resource, not the string shim. */
NEAR char m33_name[] = "FARRSC.RSC";

/* Shorter than the string it replaces ("Row 01: far tree, object   2"),
 * because menu_text's contract is that the caller made the old one long
 * enough.  NEAR for the same reason m33_name is. */
NEAR char m33_mtext[] = "MENU_TEXT WENT FAR";

/* menu_text through a FAR tree, checked at the bytes.  The object is the
 * dialog's first row, a G_STRING whose ob_spec is a plain string address.
 *
 * WATCH_SET would not do this: bank $00 is SRAM on a Rapidus and the
 * watchpoint does not see it, and a debugger breakpoint stops the sim
 * dead.  So the program reads both addresses itself, before and after,
 * and publishes two words. */
static void check_menu_text(OBJECT *tree)
{
    uint32_t spec = (uint32_t)tree[FR_ROW0].ob_spec.index;
    const char *ffar = (const char *)spec;
    const char *fnear = (const char *)(uint32_t)(uint16_t)spec;
    char before[8];
    WORD i;

    m33_mtlo = (WORD)(uint16_t)spec;
    for (i = 0; i < 8; i++)
        before[i] = fnear[i];

    menu_text(tree, FR_ROW0, m33_mtext);

    m33_mtfar = 1;
    for (i = 0; m33_mtext[i]; i++)
        if (ffar[i] != m33_mtext[i]) {
            m33_mtfar = 0;
            break;
        }
    m33_mtnear = 1;
    for (i = 0; i < 8; i++)
        if (fnear[i] != before[i]) {
            m33_mtnear = 0;
            break;
        }
}

int main(void)
{
    OBJECT *tree = 0;
    void *p;
    WORD x, y;

    m33_model = (WORD)sizeof(void *);
    m33_step = 1;                       /* entered main */
    appl_init();
    m33_step = 2;
    m33_loaded = rsrc_load(m33_name);
    m33_step = 3;
    if (m33_loaded) {
        m33_ap5 = global[5];
        m33_ap6 = global[6];
        m33_hdrlo = global[7];
        m33_hdrhi = global[8];
        if (rsrc_gaddr(R_TREE, FR_DIALOG, &p)) {
            tree = (OBJECT *)p;
            m33_bank = (WORD)((uint32_t)p >> 16);
            m33_lo = (WORD)(uint32_t)p;
        }
        m33_step = 4;
        if (tree) {
            form_center(tree, &m33_cx, &m33_cy, &m33_cw, &m33_ch);
            objc_draw(tree, ROOT, MAX_DEPTH, m33_cx, m33_cy, m33_cw, m33_ch);
            m33_step = 5;
            objc_offset(tree, FR_OK, &x, &y);
            m33_find = objc_find(tree, ROOT, MAX_DEPTH, (WORD)(x + 4), (WORD)(y + 4));
            m33_findok = (WORD)(m33_find == FR_OK);
            m33_step = 6;
        }
        if (rsrc_gaddr(R_STRING, 0, &p))
            m33_str0 = (WORD)(uint8_t)*(const char *)p;
        m33_step = 7;                   /* drawn: the gate looks now */
        evnt_keybd();
        /* AFTER the screen check, so that rewriting a row's text cannot
         * move the dark-pixel count the gate measures. */
        if (tree)
            check_menu_text(tree);
        m33_gfree = rsrc_free();
        m33_step = 8;
    }
    /* Both builds stop HERE, with everything published, until the gate
     * has read it: a program's near region is the shell's again the
     * moment it exits, and the first version of this gate read a freed
     * region and got the desktop's variables back. */
    m33_step = 9;
    evnt_keybd();
    appl_exit();
    return m33_loaded;
}

File: tests/emu/m33_farrsc.py
#!/usr/bin/env python3
"""Phase 45 gate: a resource in far memory, and the same file refused.

FARRSC.RSC is 42 KB -- 702 objects in 28 trees (tools/farrsc.py) -- and
the application pool is 14 KB of bank $00.  Two builds of one program
(src/m33_farrsc.c) ask for it under the stand-in desktop:

  F   M33.G4A, --data-model=large.  Its kit says it can take a far
      address, so rsrc_load puts the file in far memory, and the program
      draws tree 0 from there, hit-tests it, reads a free string through
      the address rsrc_gaddr handed back, and frees it.  The gate reads
      each of those results out of the program's NEAR variables, requires
      the addresses to be ABOVE bank $00 -- which is the whole claim --
      and looks at the screen while the dialog is up.

  G   M33S.G4A, --data-model=small.  Its kit passes no int_in, so it
      cannot ask, and a 16-bit program handed a far resource would have
      its pointers truncated with no error.  rsrc_load must answer 0 and
      nothing else may happen.

Both from one source, so the only difference is the model and the word
the kit sends (docs/far-trees.md, "who may receive a far address").

What the screen check is: the dialog's rectangle, as form_center placed
it, must carry text -- a count of dark pixels well above what an empty
box has.  It is not yet the pixel-exact comparison m4 makes for a pool
tree; the drawing goes through the same objc_draw, and that comparison
is the follow-up once the far tree can be handed to aesref's layout.
"""
import os
import sys

ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..")
sys.path.insert(0, os.path.join(ROOT, "tools"))
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from a8test.launcher import launch          # noqa: E402
import aesref, symfile, vbxeref             # noqa: E402
import farrsc                               # noqa: E402
from m7_form import (poke16, NOT_STARTED, STATUS, ST_GO,  # noqa: E402
                     SYMS)
from m4_aes import PRELUDE, SHOTDIR         # noqa: E402
from m12_file import Runner                 # noqa: E402
from m14_sparta import boot, screen         # noqa: E402
from m16_shell import SHELL, poll           # noqa: E402
from m17_desktop import header              # noqa: E402

DISK = os.path.abspath(os.path.join(ROOT, "build", "m33-boot.atr"))
APP = os.path.join(ROOT, "build", "m33_farrsc.g4a")
APP_SYM = os.path.join(ROOT, "build", "m33_farrsc.sym")
APPS = os.path.join(ROOT, "build", "m33s_farrsc.g4a")
APPS_SYM = os.path.join(ROOT, "build", "m33s_farrsc.sym")
RSC = os.path.join(ROOT, "build", "farrsc.rsc")
NAMES = ("m33_step", "m33_loaded", "m33_bank", "m33_lo", "m33_hdrlo", "m33_hdrhi",
         "m33_ap5", "m33_ap6", "m33_cx", "m33_cy", "m33_cw", "m33_ch",
         "m33_find", "m33_findok", "m33_str0", "m33_gfree", "m33_model",
         "m33_mtfar", "m33_mtnear", "m33_mtlo")
POOL = 14336
DARK_MIN = 2000                 # a 40x28-cell dialog with 20 rows of text has far more


def app_vars(b, syms, sym_path, g4a_path):
    """The program's NEAR variables, read by symbol against where the
    loader put its near region (the m29 pattern)."""
    link_near, near_size, far_banks = header(g4a_path)
    asym = symfile.load(sym_path)
    near = b.peek16(syms["app_near"])
    at = {n: asym[n] + near - link_near for n in NAMES}
    return {n: b.peek16(a) for n, a in at.items()}, at, near


def dark_pixels(path, x, y, w, h):
    from PIL import Image
    px = Image.open(path).convert("RGB").load()
    n = 0
    for yy in range(y, y + h):
        for xx in range(x, x + w):
            r, g, bl = px[vbxeref.SHOT_X0 + xx, vbxeref.SHOT_Y0 + yy]
            if r + g + bl < 384:
                n += 1
    return n


def main(argv=()):
    keep = "--keep" in argv
    fails = []

    def check(cond, msg):
        if not cond:
            fails.append(msg)
            print(f"  FAIL: {msg}")
        return cond

    size = os.path.getsize(RSC)
    check(size > POOL, f"the fixture is {size} bytes, which FITS the {POOL}-byte pool -- "
                       "this gate would then prove nothing")
    check(size <= 0xFFFF, f"the fixture is {size} bytes, over the one-bank limit")
    print(f"  FARRSC.RSC: {size} bytes, {farrsc.NOBS_TOTAL} objects in "
          f"{1 + farrsc.FILLER_TREES} trees; the pool is {POOL}")
    syms = symfile.load(SYMS)

    emu = launch(tag="m33", memsize="1088K", extra_args=["--disk", DISK])
    b = emu.bridge
    try:
        t, st = boot(b)
        if st is None:
            print("FAIL: the runner did not come up")
            for ln in screen(b):
                if ln.strip():
                    print("   |" + ln)
            return 1
        print(f"  M3.COM loaded and running {t} frames after RETURN")
        r = Runner(b, syms)
        r.run(PRELUDE)
        script = aesref.encode(PRELUDE + [(SHELL, (), ())], 0)
        b.memload(r.sa, b"".join(
            (x if x < 32768 else x - 65536).to_bytes(2, "little", signed=True)
            for x in script))
        poke16(b, r.count, NOT_STARTED)
        b.poke(STATUS + ST_GO, 1)
        runs = syms["sh_runs"]
        check(poll(b, runs, 1) >= 0, "the stand-in desktop never started")

        # ---- F: the large-data build, which may take a far address -------
        b.key("F")
        tf = poll(b, runs, 2)
        check(tf >= 0, f"after F: M33.G4A did not start (sh_runs {b.peek16(runs)})")
        if tf < 0:
            return 1
        v, at, near = app_vars(b, syms, APP_SYM, APP)
        tw = poll(b, at["m33_step"], 7)     # drawn, and waiting for a key
        v, at, near = app_vars(b, syms, APP_SYM, APP)
        print(f"  M33.G4A up {tf} frames after F; near ${near:04X}; "
              f"step {v['m33_step']}, {'waiting' if tw >= 0 else 'NOT waiting'} "
              f"{tw if tw >= 0 else ''}".rstrip())
        check(v["m33_model"] == 4, f"M33.G4A's pointers are {v['m33_model']} bytes; expected 4")
        check(v["m33_loaded"] == 1, f"rsrc_load returned {v['m33_loaded']}, expected 1")
        if v["m33_loaded"]:
            hdr = (v["m33_hdrhi"] << 16) | v["m33_hdrlo"]
            tree = (v["m33_bank"] << 16) | v["m33_lo"]
            ptree = (v["m33_ap6"] << 16) | v["m33_ap5"]
            print(f"  the resource at ${hdr:06X} (global[7..8]); tree 0 at ${tree:06X}; "
                  f"ap_ptree ${ptree:06X}")
            check(v["m33_hdrhi"] != 0,
                  f"the AES put the resource in bank $00 (${hdr:06X}) -- it did not go far")
            check(v["m33_bank"] != 0 and v["m33_bank"] == v["m33_hdrhi"],
                  f"tree 0 at ${tree:06X} is not in the resource's bank ${v['m33_hdrhi']:02X}")
            check(v["m33_ap6"] == v["m33_hdrhi"] and hdr < ptree < hdr + size,
                  f"ap_ptree ${ptree:06X} is not inside the resource at ${hdr:06X}")
            check(v["m33_step"] >= 7, f"the program stopped at step {v['m33_step']}")
            print(f"  form_center: ({v['m33_cx']}, {v['m33_cy']}) {v['m33_cw']}x{v['m33_ch']}; "
                  f"objc_find at OK -> {v['m33_find']} (want {farrsc.FR_OK})")
            check(v["m33_findok"] == 1,
                  f"objc_find inside the OK button answered {v['m33_find']}, expected {farrsc.FR_OK}")
            check(v["m33_str0"] == farrsc.FR_FREE0_FIRST,
                  f"free string 0 read as {v['m33_str0']!r} through its far address, "
                  f"expected {farrsc.FR_FREE0_FIRST} ('{farrsc.FREE0[0]}')")
            os.makedirs(SHOTDIR, exist_ok=True)
            shot = os.path.join(SHOTDIR, "m33-far-dialog.png")
            b.screenshot(shot)
            cx, cy, cw, chh = v["m33_cx"], v["m33_cy"], v["m33_cw"], v["m33_ch"]
            ok_rect = 0 <= cx and 0 <= cy and cw > 0 and chh > 0 and cx + cw <= 640 and cy + chh <= 240
            check(ok_rect, f"form_center gave an impossible rectangle ({cx},{cy}) {cw}x{chh}")
            if ok_rect:
                dark = dark_pixels(shot, cx, cy, cw, chh)
                print(f"  {dark} dark pixels inside the dialog's rectangle ({shot})")
                check(dark >= DARK_MIN,
                      f"only {dark} dark pixels in the dialog's rectangle: the far tree was not drawn")
            if not fails and not keep:
                os.remove(shot)
            # let it free the resource; it then waits again, at step 9, so
            # its answer can be read before its near region goes back to
            # the shell
            b.key("RETURN")
            te = poll(b, at["m33_step"], 9)
            v, _, _ = app_vars(b, syms, APP_SYM, APP)
            # menu_text through the FAR tree (src/aes/menu.c mn_text).  It
            # took (uint16_t)ob_spec until 2026-09-18, and since far_alloc
            # hands out whole banks the remainder is the string's OFFSET IN
            # THE FILE -- so the copy went to bank $00 at that offset, on
            # top of whatever lives there.  On a SpartaDOS X machine that
            # was the DOS, and the next directory read jumped into a BRK:
            # a file selector that drew and then hung, three removes from
            # the menu call that did it.
            print(f"  menu_text: wrote far {'yes' if v['m33_mtfar'] else 'NO'}, "
                  f"bank $00 at ${v['m33_mtlo']:04X} "
                  f"{'untouched' if v['m33_mtnear'] else 'CLOBBERED'}")
            check(v["m33_mtfar"] == 1,
                  "menu_text did not write where the far tree's ob_spec points")
            check(v["m33_mtnear"] == 1,
                  f"menu_text wrote into bank $00 at ${v['m33_mtlo']:04X}: "
                  f"the bank was dropped from ob_spec")
            check(te >= 0 and v["m33_gfree"] == 1,
                  f"rsrc_free returned {v['m33_gfree']} (step {v['m33_step']}), expected 1")
            b.key("RETURN")                 # and leave
        # sh_runs counts the desktop too: its return is run 3
        check(poll(b, runs, 3) >= 0, f"the desktop did not come back after M33.G4A "
                                     f"(sh_runs {b.peek16(runs)})")
        b.frames(60)

        # ---- G: the small-data build, which cannot, and must be refused ---
        b.key("G")
        tg = poll(b, runs, 4)
        check(tg >= 0, f"after G: M33S.G4A did not start (sh_runs {b.peek16(runs)})")
        if tg >= 0:
            vs, ats, nears = app_vars(b, syms, APPS_SYM, APPS)
            te = poll(b, ats["m33_step"], 9)    # refused, and waiting to be read
            vs, _, _ = app_vars(b, syms, APPS_SYM, APPS)
            print(f"  M33S.G4A up {tg} frames after G; near ${nears:04X}; "
                  f"rsrc_load returned {vs['m33_loaded']}, step {vs['m33_step']}")
            check(vs["m33_model"] == 2, f"M33S.G4A's pointers are {vs['m33_model']} bytes; expected 2")
            check(vs["m33_loaded"] == 0,
                  f"rsrc_load gave a small-data program a resource that cannot fit the pool")
            check(te >= 0, f"M33S.G4A did not reach its wait (step {vs['m33_step']})")
            b.key("RETURN")                 # and leave
    finally:
        emu.stop()

    print()
    print(f"gem4xe-m33: {'PASS' if not fails else 'FAIL'} -- a resource in far memory, "
          f"and the same file refused, {len(fails)} problem(s)")
    return 1 if fails else 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
